File: src/utils/vinMatcher.ts
// src/utils/vinMatcher.ts

export interface VehicleProfile {
  vin?: string;
  make: string;
  model: string;
  year: string;
  engine?: string;
}

// 🧠 مطابقة وتوحيد أسماء السيارات (عربي / إنجليزي)
export const normalizeVehicleText = (str: string): string => {
  return (str || '')
    .toLowerCase()
    .trim()
    .replace(/[^\w\u0621-\u064A0-9]/g, '');
};

// 🚗 قاموس المرادفات والموديلات المتوافقة
const MODEL_ALIASES: Record<string, string[]> = {
  'landcruiser': ['landcruiser', 'land cruiser', 'لاندكروزر', 'لاند كروزر', 'lc', 'lc200', 'lc300', 'v8', 'gxr', 'vxr'],
  'patrol': ['patrol', 'باترول', 'فتك', 'safari', 'nismo', 'y61', 'y62'],
  'camry': ['camry', 'كامري', 'كامري هايبرد'],
  'corolla': ['corolla', 'كورولا', 'كرولا'],
  'hilux': ['hilux', 'هايلوكس', 'هايلكس'],
  'elantra': ['elantra', 'النترا', 'إلنترا', 'avante'],
  'sonata': ['sonata', 'سوناتا'],
  'tahoe': ['tahoe', 'تاهو'],
  'yukon': ['yukon', 'يوكن', 'يوكون'],
  'prado': ['prado', 'برادو', 'لاندكروزر برادو']
};

export const isModelMatched = (dbModel: string, targetModel: string): boolean => {
  if (!dbModel || !targetModel) return true;
  const d = normalizeVehicleText(dbModel);
  const t = normalizeVehicleText(targetModel);

  if (d === t || d.includes(t) || t.includes(d)) return true;

  for (const key of Object.keys(MODEL_ALIASES)) {
    const list = MODEL_ALIASES[key].map(normalizeVehicleText);
    if (list.some(alias => t.includes(alias)) && list.some(alias => d.includes(alias))) {
      return true;
    }
  }
  return false;
};

// 📅 فحص مدى السنوات (مثال: 2018-2023 أو 2020)
export const isYearMatched = (dbYear: string | number, targetYear: string | number): boolean => {
  if (!dbYear || !targetYear) return true;
  const dbStr = String(dbYear).trim();
  const target = Number(targetYear);

  if (isNaN(target)) return true;

  if (dbStr.includes('-')) {
    const [start, end] = dbStr.split('-').map(y => Number(y.trim()));
    if (!isNaN(start) && !isNaN(end)) {
      return target >= Math.min(start, end) && target <= Math.max(start, end);
    }
  }
  return dbStr === String(targetYear) || dbStr.includes(String(targetYear));
};

/**
 * 🎯 محرك المطابقة الشامل (يدعم المطابقة التقليدية ومطابقة الإكسل برقم الشاصي)
 */
export const isPartExactFit = (part: any, vehicle: VehicleProfile): boolean => {
  if (!vehicle || (!vehicle.make && !vehicle.vin)) return true;

  // 1️⃣ [نظام الإكسل المستقبلي]: فحص عمود أرقام الشواصي المتوافقة (compatible_vins / vin_codes)
  const excelVins = part.compatible_vins || part.vin_numbers || part.vins || part.chassis_code;
  if (excelVins && vehicle.vin) {
    const cleanVin = vehicle.vin.toUpperCase().trim();
    const vinList = String(excelVins)
      .toUpperCase()
      .split(/[,;\s\n/]+/)
      .map(v => v.trim())
      .filter(Boolean);

    // مطابقة برقم الشاصي الكامل (17 حرف) أو بكود هيكل الشاصي (WMI/VDS أول 8-11 حرف)
    const hasVinMatch = vinList.some(v => cleanVin === v || cleanVin.startsWith(v) || v.startsWith(cleanVin.substring(0, 8)));
    if (hasVinMatch) return true;
  }

  // 2️⃣ مطابقة الشركة الصانعة (Make)
  const partMake = normalizeVehicleText(part.make || '');
  const vehicleMake = normalizeVehicleText(vehicle.make || '');
  const makeMatch = !partMake || !vehicleMake || partMake.includes(vehicleMake) || vehicleMake.includes(partMake);

  if (!makeMatch) return false;

  // 3️⃣ مطابقة الموديل (Model)
  const modelMatch = isModelMatched(part.model || '', vehicle.model || '');
  if (!modelMatch) return false;

  // 4️⃣ مطابقة سنة الصنع (Year)
  const yearMatch = isYearMatched(part.year || '', vehicle.year || '');
  if (!yearMatch) return false;

  // 5️⃣ مطابقة سعة المحرك (Engine) إذا توفرت
  if (vehicle.engine && part.engine && !part.engine.includes('All') && !part.engine.includes('جميع')) {
    const pEng = normalizeVehicleText(part.engine);
    const vEng = normalizeVehicleText(vehicle.engine);
    if (!pEng.includes(vEng) && !vEng.includes(pEng)) {
      return false;
    }
  }

  return true;
};

File: mobile/lib/utils/vin_matcher.dart
import '../models/vehicle_model.dart';

typedef SelectedVehicle = VehicleProfile;

class VinMatcher {
  static String normalizeVehicleText(String str) {
    return (str)
        .toLowerCase()
        .trim()
        .replaceAll(RegExp(r'[^\w\u0621-\u064A0-9]'), '');
  }

  static const Map<String, List<String>> modelAliases = {
    'landcruiser': [
      'landcruiser',
      'land cruiser',
      'لاندكروزر',
      'لاند كروزر',
      'lc',
      'lc200',
      'lc300',
      'v8',
      'gxr',
      'vxr',
    ],
    'patrol': ['patrol', 'باترول', 'فتك', 'safari', 'nismo', 'y61', 'y62'],
    'camry': ['camry', 'كامري', 'كامري هايبرد'],
    'corolla': ['corolla', 'كورولا', 'كرولا'],
    'hilux': ['hilux', 'هايلوكس', 'هايلكس'],
    'elantra': ['elantra', 'النترا', 'إلنترا', 'avante'],
    'sonata': ['sonata', 'سوناتا'],
    'tahoe': ['tahoe', 'تاهو'],
    'yukon': ['yukon', 'يوكن', 'يوكون'],
    'prado': ['prado', 'برادو', 'لاندكروزر برادو'],
    'optima': ['optima', 'k5', 'أوبتيما', 'اوبتيما', 'اوبتيما / k5'],
    'taurus': ['taurus', 'تورس', 'توروس'],
    'altima': ['altima', 'التيما', 'ألتيما'],
    'accord': ['accord', 'اكورد', 'أكورد'],
    'eclass': ['e-class', 'e class', 'e300', 'e200', 'e350', 'e class'],
  };

  /// Matches DB model against selected/target model (web: isModelMatched(dbModel, targetModel)).
  static bool isModelMatched(String dbModel, String targetModel) {
    if (dbModel.isEmpty || targetModel.isEmpty) return true;
    final d = normalizeVehicleText(dbModel);
    final t = normalizeVehicleText(targetModel);

    if (d == t || d.contains(t) || t.contains(d)) return true;

    for (final aliases in modelAliases.values) {
      final list = aliases.map(normalizeVehicleText).toList();
      if (list.any((a) => t.contains(a)) && list.any((a) => d.contains(a))) {
        return true;
      }
    }
    return false;
  }

  /// Matches DB year (single or range "2018-2023") against selected year.
  static bool isYearMatched(dynamic dbYear, dynamic targetYear) {
    if (dbYear == null || targetYear == null) return true;
    final dbStr = dbYear.toString().trim();
    final targetStr = targetYear.toString().trim();
    if (dbStr.isEmpty || targetStr.isEmpty) return true;

    final target = int.tryParse(targetStr);
    if (target == null) return true;

    if (dbStr.contains('-')) {
      final parts = dbStr.split('-').map((y) => int.tryParse(y.trim())).toList();
      if (parts.length == 2 && parts[0] != null && parts[1] != null) {
        final start = parts[0]!;
        final end = parts[1]!;
        return target >= (start < end ? start : end) &&
            target <= (start > end ? start : end);
      }
    }
    return dbStr == targetStr || dbStr.contains(targetStr);
  }

  static bool isMakeMatched(String dbMake, String targetMake) {
    if (dbMake.isEmpty || targetMake.isEmpty) return true;
    final d = normalizeVehicleText(dbMake);
    final t = normalizeVehicleText(targetMake);
    return d.contains(t) || t.contains(d);
  }

  static bool isPartExactFit(Map<String, dynamic> part, VehicleProfile vehicle) {
    if (vehicle.make.isEmpty && (vehicle.vin == null || vehicle.vin!.isEmpty)) {
      return true;
    }

    final excelVins = part['compatible_vins'] ??
        part['vin_numbers'] ??
        part['vins'] ??
        part['chassis_code'];
    if (excelVins != null && vehicle.vin != null && vehicle.vin!.isNotEmpty) {
      final cleanVin = vehicle.vin!.toUpperCase().trim();
      final vinList = excelVins
          .toString()
          .toUpperCase()
          .split(RegExp(r'[,;\s\n/]+'))
          .map((v) => v.trim())
          .where((v) => v.isNotEmpty);
      final hasVinMatch = vinList.any(
        (v) =>
            cleanVin == v ||
            cleanVin.startsWith(v) ||
            v.startsWith(cleanVin.substring(0, cleanVin.length.clamp(0, 8))),
      );
      if (hasVinMatch) return true;
    }

    final partMake = (part['make'] ?? '').toString();
    if (!isMakeMatched(partMake, vehicle.make)) return false;
    if (!isModelMatched((part['model'] ?? '').toString(), vehicle.model)) {
      return false;
    }
    if (!isYearMatched(part['year'], vehicle.year)) return false;

    if (vehicle.engine != null &&
        vehicle.engine!.isNotEmpty &&
        part['engine'] != null) {
      final pEng = part['engine'].toString();
      if (!pEng.contains('All') && !pEng.contains('جميع')) {
        final pe = normalizeVehicleText(pEng);
        final ve = normalizeVehicleText(vehicle.engine!);
        if (!pe.contains(ve) && !ve.contains(pe)) return false;
      }
    }
    return true;
  }

  static VehicleProfile? matchVin(String vin) {
    final cleanVin = vin.trim().toUpperCase();
    if (cleanVin.length != 17) return null;
    if (cleanVin.startsWith('J')) {
      return VehicleProfile(
        vin: cleanVin,
        make: 'تويوتا',
        model: 'لاندكروزر',
        year: '2023',
      );
    }
    return null;
  }
}
